File: Problems/Algorithms/1335. Minimum Difficulty of a Job Schedule/MinDifficultySchedule.java
class Solution {
    public int minDifficulty(int[] jobDifficulty, int d) {
        int n = jobDifficulty.length;

        if (n < d) {
            return -1;
        }

        int[][] dp = new int[d+1][n+1];
        for (int i = 0; i <= d; i++) {
            Arrays.fill(dp[i], Integer.MAX_VALUE);
        }
        dp[0][0] = 0;
        for (int i = 1; i <= d; i++) {
            for (int j = i; j <= n; j++) {
                int max = 0;
                for (int k = j; k >= i; k--) {
                    max = Math.max(max, jobDifficulty[k-1]);
                    if (dp[i-1][k-1] != Integer.MAX_VALUE) {
                        dp[i][j] = Math.min(dp[i][j], dp[i-1][k-1] + max);
                    }
                }
            }
        }

        return dp[d][n];
    }
}

File: Problems/Algorithms/1335. Minimum Difficulty of a Job Schedule/min_difficulty_schedule.py
#!/usr/bin/env python3

from functools import lru_cache
from typing import List
from math import inf


class Solution:
    def minDifficulty(self, jobDifficulty: List[int], d: int) -> int:
        dp = {}

        @lru_cache
        def helper(idx, d, maximum):
            if (idx, d, maximum) in dp:
                return dp[(idx, d, maximum)]
            if idx == len(jobDifficulty) and d == 0:
                return maximum
            if idx == len(jobDifficulty) or d == 0:
                return float(inf)

            maximum = max(maximum, jobDifficulty[idx])
            ans = min(helper(idx+1, d, maximum), maximum + helper(idx+1, d-1, 0))
            dp[(idx, d, maximum)] = ans
            return ans

        ans = helper(0, d, 0)
        return ans if ans != float(inf) else -1
